src/design_notes/general_coroutines.md

```diff
@@ -0,0 +1,258 @@
+# Generalized coroutines
+
+Since even before Rust 1.0, users have desired the ability to `yield` like in
+other languages. The compiler infrastructure to achieve this, along with an
+unstable syntax, have existed for a while now. But despite *a lot* of debate,
+we've failed to polish the feature up enough to stabilize it. I've tried to
+write up a summary of the different design considerations and the past debate
+around them below:
+
+## Coroutines vs Generators
+
+- The distinction between a "coroutine" and a "generator" can be a bit vague,
+  varying from one discussion to the next.
+- In these notes a generator is anything which *directly* implements `Iterator`
+  or `Stream` while a coroutine is anything which can take arbitrary input,
+  yield arbitrary output, and later resume execution at the previous yield.
+- Thus, the "generator" syntax proposed in [eRFC-2033][1] and currently
+  implemented behind the "generator" feature is actually a coroutine syntax for
+  the sake of these notes, *not a true generator*.
+- Note also that "coroutines" here are really "semicoroutines" since they can
+  only yield back to their caller.
+- Generators are a specialization of coroutines.
+- I will continue to group the [original eRFC text][1] and the later [generator
+  resume arguments](https://github.com/rust-lang/rust/pull/68524) extension
+  togther as "[eRFC-2033][1]". That way I only have 3 big proposals to deal
+  with.
+
+```rust
+// This is a coroutine
+|stuff| {
+  yield stuff + 1;
+  yield stuff + 2;
+  return stuff + 3;
+}
+
+// This is a generator
+stream! {
+  yield 1;
+  yield 2;
+  yield 3;
+}
+```
+
+## Coroutine trait
+
+- The coroutine syntax must produce implementations of some trait.
+- [RFC-2781][2] and [eRFC-2033][1] propose the `Generator` trait.
+- Note that Rust's coroutines and subroutines look the same from the outside:
+  take input, mutate state, produce output.
+- Thus, [MCP-49][3] proposes using the `Fn*` traits instead, including a new
```

Comment:
This is still not enough to express attached/lending/whatever-we're calling them (I keep forgetting) structures, as the `Fn` traits don't allow for returning items that borrow from the closure state itself (though we want that feature for other reasons, chiefly async closures).

Reply:
Correct. I'll add a little section on lending to clarify this.

Since we eventually want lending closures anyway, MCP-49 treats it as an orthogonal problem: once we get lending closures, futures/streams/etc would be able to lend as well. I believe RFC-2781 also treats it as orthogonal since the RFC text never mentions lending and does not use GATs for `Yield` and `Return` types.

Reply:
Lending!